lib/free_quota.rb

```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+require "yaml"
+
+class FreeQuota
+  # :nocov:
+  def self.freeze
+    free_quotas
+    super
+  end
+  # :nocov:
+
+  def self.free_quotas
+    @free_quotas ||= begin
+      quotas = YAML.load_file("config/free_quotas.yml")
+      quotas.each_with_object({}) do |item, hash|
+        hash[item["name"]] = item
+      end
+    end
+    @free_quotas
+  end
+
+  def self.get_applicable_billing_rate_ids(free_quota)
+    BillingRate.from_resource_type(free_quota["resource_type"]).map { _1["id"] }
+  end
+
+  def self.remaining_free_quota(name, project_id)
+    free_quota = free_quotas[name]
+    begin_time = get_begin_time
+    tokens_used = BillingRecord.where(
+      project_id: project_id,
+      billing_rate_id: get_applicable_billing_rate_ids(free_quota)
+    ).where {
+      Sequel.pg_range(span).overlaps(Sequel.pg_range(begin_time...Time.now))
+    }.sum(:amount) || 0
+    [0, free_quota["value"] - tokens_used].max
+  end
+
+  def self.get_free_quota_exhausted_projects(name)
```

Comment:
It returns a dataset, and also we call it via `FreeQuota.get_free_quota_exhausted_projects`, no need to say free quota twice.

`FreeQuota.exhausted_projects_ds(name)` looks more natural to me.

```suggestion
  def self.exhausted_projects_ds(name)
```

Reply:
Done.